In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings


warnings.filterwarnings('ignore')
%matplotlib inline

C:\Anaconda\lib\site-packages\scipy\__init__.py:138: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.4)
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion} is required for this version of "


In [2]:
#!C:\Anaconda\python.exe -m pip install --upgrade pip

In [3]:
pd.__version__

'1.3.2'

In [4]:
#!pip install pandas==1.3.2
#!pip install vaex

In [5]:
#meta = vaex.to

In [6]:
meta = pd.read_csv('movies_metadata.csv', nrows=5000)
keywords = pd.read_csv('keywords.csv', nrows=5000)
credits = pd.read_csv('credits.csv', nrows=5000)

In [7]:
meta.head(3)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   

     original_title                                           overview  ...  \
0         Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1           Jumanji  When siblings Judy and Peter discover an encha...  ...   
2  Grumpier Old Men  A family wedding reignites the ancient feud be...  ...   

   release_date    revenue runtime  \
0    1995-10-30  373554033    81.0   
1    1995-12-15  262797249   104.0   
2    1995-12-22          0   101.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

  vote_average vote_count  
0          7.7       5415  
1          6.9       2413  
2          6.5         92  

[3 rows x 24 columns]

In [8]:
meta.shape

(5000, 24)

In [9]:
meta.isnull().sum()

adult                       0
belongs_to_collection    4175
budget                      0
genres                      0
homepage                 4689
id                          0
imdb_id                     0
original_language           0
original_title              0
overview                   21
popularity                  0
poster_path                21
production_companies        0
production_countries        0
release_date                4
revenue                     0
runtime                     6
spoken_languages            0
status                      6
tagline                  1251
title                       0
video                       0
vote_average                0
vote_count                  0
dtype: int64

We have date entries in our id column.

In [10]:
meta[meta['id'] == '1997-08-20']

Empty DataFrame
Columns: [adult, belongs_to_collection, budget, genres, homepage, id, imdb_id, original_language, original_title, overview, popularity, poster_path, production_companies, production_countries, release_date, revenue, runtime, spoken_languages, status, tagline, title, video, vote_average, vote_count]
Index: []

[0 rows x 24 columns]

Let's check all rows having date entries in id column

In [11]:
# meta[meta['id'].str.contains('-') == True].index

In [12]:
#meta.drop([19730, 29503], axis=0, inplace=True)

In [13]:
#meta['id'] = meta['id'].astype('int64')
meta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  5000 non-null   bool   
 1   belongs_to_collection  825 non-null    object 
 2   budget                 5000 non-null   int64  
 3   genres                 5000 non-null   object 
 4   homepage               311 non-null    object 
 5   id                     5000 non-null   int64  
 6   imdb_id                5000 non-null   object 
 7   original_language      5000 non-null   object 
 8   original_title         5000 non-null   object 
 9   overview               4979 non-null   object 
 10  popularity             5000 non-null   float64
 11  poster_path            4979 non-null   object 
 12  production_companies   5000 non-null   object 
 13  production_countries   5000 non-null   object 
 14  release_date           4996 non-null   object 
 15  reve

In [14]:
meta['poster_path']

0       /rhIRbceoE9lR4veEXuwCC2wARtG.jpg
1       /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg
2       /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg
3       /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg
4       /e64sOI48hQXyru7naBFyssKFxVd.jpg
                      ...               
4995    /k7iUnaelPU5kwEEVZV7xW7fvFIS.jpg
4996    /gs3fFBoiEaExy5x3m5yAQpmVhXs.jpg
4997    /jSNMRMPkXUqcqWm0xUPUviRfWfF.jpg
4998     /iDkgvzFjIvs7n0C4gDgAh7hnSY.jpg
4999    /4qNxipsoqADZu5ZACK6uJv4qkWV.jpg
Name: poster_path, Length: 5000, dtype: object

In [15]:
keywords.head(3)

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...

In [16]:
keywords['keywords'][0]

"[{'id': 931, 'name': 'jealousy'}, {'id': 4290, 'name': 'toy'}, {'id': 5202, 'name': 'boy'}, {'id': 6054, 'name': 'friendship'}, {'id': 9713, 'name': 'friends'}, {'id': 9823, 'name': 'rivalry'}, {'id': 165503, 'name': 'boy next door'}, {'id': 170722, 'name': 'new toy'}, {'id': 187065, 'name': 'toy comes to life'}]"

In [17]:
keywords.shape

(5000, 2)

In [18]:
keywords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        5000 non-null   int64 
 1   keywords  5000 non-null   object
dtypes: int64(1), object(1)
memory usage: 78.2+ KB


In [19]:
credits.head(3)

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602

In [20]:
credits.shape

(5000, 3)

In [21]:
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    5000 non-null   object
 1   crew    5000 non-null   object
 2   id      5000 non-null   int64 
dtypes: int64(1), object(2)
memory usage: 117.3+ KB


In [22]:
df = pd.merge(meta, keywords, on='id')
df.head(3)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   

     original_title                                           overview  ...  \
0         Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1           Jumanji  When siblings Judy and Peter discover an encha...  ...   
2  Grumpier Old Men  A family wedding reignites the ancient feud be...  ...   

     revenue runtime                                   spoken_languages  \
0  373554033    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   

              title  video vote_average vote_count  \
0         Toy Story  False          7.7       5415   
1           Jumanji  False          6.9       2413   
2  Grumpier Old Men  False          6.5         92   

                                            keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...  

[3 rows x 25 columns]

In [23]:
df = pd.merge(df, credits, on='id')
df.head(2)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  ...  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1        Jumanji  When siblings Judy and Peter discover an encha...  ...   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   

                                     tagline      title  video  vote_average  \
0                                        NaN  Toy Story  False           7.7   
1  Roll the dice and unleash the excitement!    Jumanji  False           6.9   

   vote_count                                           keywords  \
0        5415  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1        2413  [{'id': 10090, 'name': 'board game'}, {'id': 1...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...  

[2 rows x 27 columns]

In [24]:
df.shape

(5006, 27)

In [25]:
df.isna().sum()

adult                       0
belongs_to_collection    4181
budget                      0
genres                      0
homepage                 4695
id                          0
imdb_id                     0
original_language           0
original_title              0
overview                   21
popularity                  0
poster_path                21
production_companies        0
production_countries        0
release_date                4
revenue                     0
runtime                     6
spoken_languages            0
status                      6
tagline                  1251
title                       0
video                       0
vote_average                0
vote_count                  0
keywords                    0
cast                        0
crew                        0
dtype: int64

In [26]:
df['original_language'].value_counts()

en    4476
fr     169
de      74
it      66
es      39
ja      38
zh      26
cn      21
sv      11
fa      10
ru      10
pt       9
nl       8
da       7
no       4
sr       4
cs       4
xx       3
pl       2
vi       2
bn       2
hu       2
th       2
he       2
ko       2
fi       1
ca       1
sk       1
bo       1
mk       1
ro       1
sh       1
el       1
nb       1
cy       1
wo       1
ar       1
bs       1
Name: original_language, dtype: int64

In [27]:
df.iloc[23:30]['spoken_languages']

23             [{'iso_639_1': 'en', 'name': 'English'}]
24             [{'iso_639_1': 'en', 'name': 'English'}]
25             [{'iso_639_1': 'en', 'name': 'English'}]
26             [{'iso_639_1': 'en', 'name': 'English'}]
27             [{'iso_639_1': 'en', 'name': 'English'}]
28    [{'iso_639_1': 'cn', 'name': '广州话 / 廣州話'}, {'i...
29                 [{'iso_639_1': 'zh', 'name': '普通话'}]
Name: spoken_languages, dtype: object

In [28]:
# Genres, original language, original title, overview, production_companies, keywords, cast, crew

In [29]:
df = df[['genres','id','poster_path','original_title','overview','production_companies','keywords','cast','crew']]
df.head()

genres     id  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357   
4                     [{'id': 35, 'name': 'Comedy'}]  11862   

                        poster_path               original_title  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg                    Toy Story   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg                      Jumanji   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg             Grumpier Old Men   
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg            Waiting to Exhale   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg  Father of the Bride Part II   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                            keywords  \
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...   
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...   
3  [{'id': 818, 'name': 'based on novel'}, {'id':...   
4  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...

In [30]:
df.shape

(5006, 9)

In [31]:
df.isnull().sum()/len(df) *100

genres                  0.000000
id                      0.000000
poster_path             0.419497
original_title          0.000000
overview                0.419497
production_companies    0.000000
keywords                0.000000
cast                    0.000000
crew                    0.000000
dtype: float64

Since, 0.4% rows don't have an overview and poster_path which is an important parameter, so we can drop them

In [32]:
df.dropna(inplace=True)

In [33]:
df.duplicated().sum()

7

In [34]:
df.drop_duplicates(inplace=True)

In [35]:
df.duplicated().sum()

0

In [36]:
df.shape

(4960, 9)

In [37]:
# English Titles only
df.original_title.replace({r'[^\x00-\x7F]+':''}, regex=True)

0                         Toy Story
1                           Jumanji
2                  Grumpier Old Men
3                 Waiting to Exhale
4       Father of the Bride Part II
                   ...             
5001              The Deadly Mantis
5002                      Dragonfly
5003            Queen of the Damned
5004                   Big Bad Love
5005                   Green Dragon
Name: original_title, Length: 4960, dtype: object

In [38]:
# Fetching rows that don't contain english words in original_title column
df[~df['original_title'].map(lambda x: x.isascii())]

genres     id  \
28    [{'id': 14, 'name': 'Fantasy'}, {'id': 878, 'n...    902   
29    [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...  37557   
72    [{'id': 18, 'name': 'Drama'}, {'id': 36, 'name...  48750   
79    [{'id': 10751, 'name': 'Family'}, {'id': 18, '...  46785   
110   [{'id': 80, 'name': 'Crime'}, {'id': 28, 'name...  33542   
...                                                 ...    ...   
4904  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...  24166   
4932  [{'id': 18, 'name': 'Drama'}, {'id': 36, 'name...  11710   
4944                      [{'id': 18, 'name': 'Drama'}]  12659   
4946                      [{'id': 18, 'name': 'Drama'}]  35650   
4962  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   4267   

                           poster_path                      original_title  \
28    /eVo6ewq4akfyJYy3GXkMsLNzEJc.jpg          La Cité des Enfants Perdus   
29    /qcoOCoN7viOhboGwhYXyApdDuiq.jpg                           摇啊摇，摇到外婆桥   
72    /ePJCk8puXx28CtvpSpmezpwfVt8.jpg                      Les misérables   
79    /96UDOJNgJXHmQMJqA79bKS9PR8i.jpg                         بادکنک سفید   
110   /y7AxKPPCtZyrGQKvxzn5zTQ8wFm.jpg                                 紅番區   
...                                ...                                 ...   
4904  /uR4OhZa8tef9jPSdBpcOSnufPa0.jpg                               你那邊幾點   
4932  /jkbnlcex3xfndksKuwVeI0EmuhW.jpg  Jeder für sich und Gott gegen alle   
4944  /tWURu1bBoBHn92hZGgG0y6JEVHU.jpg                              十七歲的單車   
4946  /ouM7H16aKd3IzvjHNcUmzezI5qq.jpg                           Maelström   
4962  /5klW5Je4RoNBpsZY6dwqdZthrzv.jpg              Préparez vos mouchoirs   

                                               overview  \
28    A scientist in a surrealist society kidnaps ch...   
29    A provincial boy related to a Shanghai crime f...   
72    In France during World War II, a poor and illi...   
79    Several people try to take advantage of a litt...   
110   Keong comes from Hong Kong to visit New York f...   
...                                                 ...   
4904  When a young street vendor with a grim home li...   
4932  The film follows Kaspar Hauser (Bruno S.), who...   
4944  A seventeen-year-old country boy working in Be...   
4946  Bibiane is the daughter of a well-known fashio...   
4962  Solange is depressed: she's stopped smiling, s...   

                                   production_companies  \
28    [{'name': 'Procirep', 'id': 311}, {'name': 'Co...   
29    [{'name': 'Ministère des Affaires Étrangères',...   
72    [{'name': 'Les Films 13', 'id': 1742}, {'name'...   
79    [{'name': 'I.R.I.B. Channel 2', 'id': 38953}, ...   
110   [{'name': 'New Line Cinema', 'id': 12}, {'name...   
...                                                 ...   
4904  [{'name': 'Arena Films', 'id': 1243}, {'name':...   
4932  [{'name': 'Werner Herzog Filmproduktion', 'id'...   
4944  [{'name': 'Arc Light Films', 'id': 7475}, {'na...   
4946  [{'name': 'Téléfilm Canada', 'id': 806}, {'nam...   
4962  [{'name': 'CAPAC', 'id': 1122}, {'name': 'Belg...   

                                               keywords  \
28    [{'id': 402, 'name': 'clone'}, {'id': 1566, 'n...   
29    [{'id': 964, 'name': 'servant'}, {'id': 2311, ...   
72    [{'id': 90, 'name': 'paris'}, {'id': 745, 'nam...   
79    [{'id': 1357, 'name': 'fish'}, {'id': 10594, '...   
110   [{'id': 242, 'name': 'new york'}, {'id': 779, ...   
...                                                 ...   
4904                                                 []   
4932  [{'id': 2434, 'name': 'liberation'}, {'id': 44...   
4944  [{'id': 642, 'name': 'robbery'}, {'id': 1850, ...   
4946        [{'id': 10183, 'name': 'independent film'}]   
4962  [{'id': 1157, 'name': 'wife husband relationsh...   

                                                   cast  \
28    [{'cast_id': 24, 'character': 'One', 'credit_i...   
29    [{'cast_id': 2, 'character': 'Xiao Jingbao', '...   
72    [{'cast_i

In [39]:
df = df[df['original_title'].map(lambda x: x.isascii())]
df.head()

genres     id  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357   
4                     [{'id': 35, 'name': 'Comedy'}]  11862   

                        poster_path               original_title  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg                    Toy Story   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg                      Jumanji   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg             Grumpier Old Men   
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg            Waiting to Exhale   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg  Father of the Bride Part II   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                            keywords  \
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...   
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...   
3  [{'id': 818, 'name': 'based on novel'}, {'id':...   
4  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...

In [40]:
df.shape

(4763, 9)

In [41]:
df

genres     id  \
0     [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1     [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2     [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   
3     [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357   
4                        [{'id': 35, 'name': 'Comedy'}]  11862   
...                                                 ...    ...   
5001  [{'id': 53, 'name': 'Thriller'}, {'id': 27, 'n...  43715   
5002                      [{'id': 18, 'name': 'Drama'}]  10052   
5003  [{'id': 18, 'name': 'Drama'}, {'id': 14, 'name...  11979   
5004  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...  75151   
5005  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...  35868   

                           poster_path               original_title  \
0     /rhIRbceoE9lR4veEXuwCC2wARtG.jpg                    Toy Story   
1     /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg                      Jumanji   
2     /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg             Grumpier Old Men   
3     /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg            Waiting to Exhale   
4     /e64sOI48hQXyru7naBFyssKFxVd.jpg  Father of the Bride Part II   
...                                ...                          ...   
5001  /k7iUnaelPU5kwEEVZV7xW7fvFIS.jpg            The Deadly Mantis   
5002  /gs3fFBoiEaExy5x3m5yAQpmVhXs.jpg                    Dragonfly   
5003  /jSNMRMPkXUqcqWm0xUPUviRfWfF.jpg          Queen of the Damned   
5004   /iDkgvzFjIvs7n0C4gDgAh7hnSY.jpg                 Big Bad Love   
5005  /4qNxipsoqADZu5ZACK6uJv4qkWV.jpg                 Green Dragon   

                                               overview  \
0     Led by Woody, Andy's toys live happily in his ...   
1     When siblings Judy and Peter discover an encha...   
2     A family wedding reignites the ancient feud be...   
3     Cheated on, mistreated and stepped on, the wom...   
4     Just when George Banks has recovered from his ...   
...                                                 ...   
5001  The calving of an Arctic iceberg releases a gi...   
5002  A grieving doctor is being contacted by his la...   
5003  Lestat de Lioncourt is awakened from his slumb...   
5004  Vietnam veteran Leon Barlow is struggling as a...   
5005  A tale about Vietnamese refugees sent to an or...   

                                   production_companies  \
0        [{'name': 'Pixar Animation Studios', 'id': 3}]   
1     [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2     [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3     [{'name': 'Twentieth Century Fox Film Corporat...   
4     [{'name': 'Sandollar Productions', 'id': 5842}...   
...                                                 ...   
5001  [{'name': 'Universal International Pictures (U...   
5002  [{'name': 'Universal Pictures', 'id': 33}, {'n...   
5003  [{'name': 'Village Roadshow Pictures', 'id': 7...   
5004                                                 []   
5005                                                 []   

                                               keywords  \
0     [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1     [{'id': 10090, 'name': 'board game'}, {'id': 1...   
2     [{'id': 1495, 'name': 'fishing'}, {'id': 12392...   
3     [{'id': 818, 'name': 'based on novel'}, {'id':...   
4     [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...   
...                                                 ...   
5001  [{'id': 1299, 'name': 'monster'}, {'id': 1719,...   
5002  [{'id': 3725, 'name': 'pregnancy and birth'}, ...   
5003  [{'id': 2011, 'name': 'queen'}, {'id': 2570, '...   
5004                                                 []   
5005        [{'id': 10183, 'name': 'independent film'}]   

                                                   cast  \
0     [{'cast_id': 14, 'character': 'Woody (voice)',...   
1     [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2     [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3     [{'cast_id': 1, 'character': "Sava

In [42]:
df.reset_index(inplace=True)
df.drop(['index'], axis=1, inplace=True)
df

genres     id  \
0     [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1     [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2     [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   
3     [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357   
4                        [{'id': 35, 'name': 'Comedy'}]  11862   
...                                                 ...    ...   
4758  [{'id': 53, 'name': 'Thriller'}, {'id': 27, 'n...  43715   
4759                      [{'id': 18, 'name': 'Drama'}]  10052   
4760  [{'id': 18, 'name': 'Drama'}, {'id': 14, 'name...  11979   
4761  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...  75151   
4762  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...  35868   

                           poster_path               original_title  \
0     /rhIRbceoE9lR4veEXuwCC2wARtG.jpg                    Toy Story   
1     /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg                      Jumanji   
2     /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg             Grumpier Old Men   
3     /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg            Waiting to Exhale   
4     /e64sOI48hQXyru7naBFyssKFxVd.jpg  Father of the Bride Part II   
...                                ...                          ...   
4758  /k7iUnaelPU5kwEEVZV7xW7fvFIS.jpg            The Deadly Mantis   
4759  /gs3fFBoiEaExy5x3m5yAQpmVhXs.jpg                    Dragonfly   
4760  /jSNMRMPkXUqcqWm0xUPUviRfWfF.jpg          Queen of the Damned   
4761   /iDkgvzFjIvs7n0C4gDgAh7hnSY.jpg                 Big Bad Love   
4762  /4qNxipsoqADZu5ZACK6uJv4qkWV.jpg                 Green Dragon   

                                               overview  \
0     Led by Woody, Andy's toys live happily in his ...   
1     When siblings Judy and Peter discover an encha...   
2     A family wedding reignites the ancient feud be...   
3     Cheated on, mistreated and stepped on, the wom...   
4     Just when George Banks has recovered from his ...   
...                                                 ...   
4758  The calving of an Arctic iceberg releases a gi...   
4759  A grieving doctor is being contacted by his la...   
4760  Lestat de Lioncourt is awakened from his slumb...   
4761  Vietnam veteran Leon Barlow is struggling as a...   
4762  A tale about Vietnamese refugees sent to an or...   

                                   production_companies  \
0        [{'name': 'Pixar Animation Studios', 'id': 3}]   
1     [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2     [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3     [{'name': 'Twentieth Century Fox Film Corporat...   
4     [{'name': 'Sandollar Productions', 'id': 5842}...   
...                                                 ...   
4758  [{'name': 'Universal International Pictures (U...   
4759  [{'name': 'Universal Pictures', 'id': 33}, {'n...   
4760  [{'name': 'Village Roadshow Pictures', 'id': 7...   
4761                                                 []   
4762                                                 []   

                                               keywords  \
0     [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1     [{'id': 10090, 'name': 'board game'}, {'id': 1...   
2     [{'id': 1495, 'name': 'fishing'}, {'id': 12392...   
3     [{'id': 818, 'name': 'based on novel'}, {'id':...   
4     [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...   
...                                                 ...   
4758  [{'id': 1299, 'name': 'monster'}, {'id': 1719,...   
4759  [{'id': 3725, 'name': 'pregnancy and birth'}, ...   
4760  [{'id': 2011, 'name': 'queen'}, {'id': 2570, '...   
4761                                                 []   
4762        [{'id': 10183, 'name': 'independent film'}]   

                                                   cast  \
0     [{'cast_id': 14, 'character': 'Woody (voice)',...   
1     [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2     [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3     [{'cast_id': 1, 'character': "Sava

In [43]:
df['genres'][0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [44]:
# We want a list as ['Animation','Comedy','Family']. 
# But as our dictionary above is a string not a list we cannot loop through it.
# We have to convert this string of list of dictionary into a list of dictionary which can be done using ast.literal_eval function

In [45]:
import ast
ast.literal_eval("[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]")

[{'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 10751, 'name': 'Family'}]

In [46]:
import ast

def get_list(obj):
    my_list = []
    for i in ast.literal_eval(obj):       #literal_eval function converts the obj in string format to a list format
        my_list.append(i['name'])
    return my_list

In [47]:
df['genres'] = df['genres'].apply(get_list)
df.head(2)

genres    id                       poster_path  \
0   [Animation, Comedy, Family]   862  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  [Adventure, Fantasy, Family]  8844  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
1        Jumanji  When siblings Judy and Peter discover an encha...   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   

                                            keywords  \
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...

In [48]:
df['keywords'][0]

"[{'id': 931, 'name': 'jealousy'}, {'id': 4290, 'name': 'toy'}, {'id': 5202, 'name': 'boy'}, {'id': 6054, 'name': 'friendship'}, {'id': 9713, 'name': 'friends'}, {'id': 9823, 'name': 'rivalry'}, {'id': 165503, 'name': 'boy next door'}, {'id': 170722, 'name': 'new toy'}, {'id': 187065, 'name': 'toy comes to life'}]"

In [49]:
df['production_companies'] = df['production_companies'].apply(get_list)
df['keywords'] = df['keywords'].apply(get_list)
df.head(3)

genres     id                       poster_path  \
0   [Animation, Comedy, Family]    862  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  [Adventure, Fantasy, Family]   8844  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2             [Romance, Comedy]  15602  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   

     original_title                                           overview  \
0         Toy Story  Led by Woody, Andy's toys live happily in his ...   
1           Jumanji  When siblings Judy and Peter discover an encha...   
2  Grumpier Old Men  A family wedding reignites the ancient feud be...   

                                production_companies  \
0                          [Pixar Animation Studios]   
1  [TriStar Pictures, Teitler Film, Interscope Co...   
2                     [Warner Bros., Lancaster Gate]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   
2  [fishing, best friend, duringcreditsstinger, o...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   

                                                crew  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...

In [50]:
df['cast'][0]

"[{'cast_id': 14, 'character': 'Woody (voice)', 'credit_id': '52fe4284c3a36847f8024f95', 'gender': 2, 'id': 31, 'name': 'Tom Hanks', 'order': 0, 'profile_path': '/pQFoyx7rp09CJTAb932F2g8Nlho.jpg'}, {'cast_id': 15, 'character': 'Buzz Lightyear (voice)', 'credit_id': '52fe4284c3a36847f8024f99', 'gender': 2, 'id': 12898, 'name': 'Tim Allen', 'order': 1, 'profile_path': '/uX2xVf6pMmPepxnvFWyBtjexzgY.jpg'}, {'cast_id': 16, 'character': 'Mr. Potato Head (voice)', 'credit_id': '52fe4284c3a36847f8024f9d', 'gender': 2, 'id': 7167, 'name': 'Don Rickles', 'order': 2, 'profile_path': '/h5BcaDMPRVLHLDzbQavec4xfSdt.jpg'}, {'cast_id': 17, 'character': 'Slinky Dog (voice)', 'credit_id': '52fe4284c3a36847f8024fa1', 'gender': 2, 'id': 12899, 'name': 'Jim Varney', 'order': 3, 'profile_path': '/eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg'}, {'cast_id': 18, 'character': 'Rex (voice)', 'credit_id': '52fe4284c3a36847f8024fa5', 'gender': 2, 'id': 12900, 'name': 'Wallace Shawn', 'order': 4, 'profile_path': '/oGE6JqPP2xH4t

We will only fetch the first 3 names in the cast column as they are probably important actors.

In [51]:
def get_actors(obj):
    act = []
    count = 0
    for i in ast.literal_eval(obj):
        if count < 3:
            act.append(i['name'])
            count += 1
        else:
            break
    return act

In [52]:
df['cast'] = df['cast'].apply(get_actors)
df.head(2)

genres    id                       poster_path  \
0   [Animation, Comedy, Family]   862  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  [Adventure, Fantasy, Family]  8844  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
1        Jumanji  When siblings Judy and Peter discover an encha...   

                                production_companies  \
0                          [Pixar Animation Studios]   
1  [TriStar Pictures, Teitler Film, Interscope Co...   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   

                                             cast  \
0             [Tom Hanks, Tim Allen, Don Rickles]   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   

                                                crew  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...

In [53]:
df['crew'][0]

'[{\'credit_id\': \'52fe4284c3a36847f8024f49\', \'department\': \'Directing\', \'gender\': 2, \'id\': 7879, \'job\': \'Director\', \'name\': \'John Lasseter\', \'profile_path\': \'/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f4f\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12891, \'job\': \'Screenplay\', \'name\': \'Joss Whedon\', \'profile_path\': \'/dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f55\', \'department\': \'Writing\', \'gender\': 2, \'id\': 7, \'job\': \'Screenplay\', \'name\': \'Andrew Stanton\', \'profile_path\': \'/pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f5b\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12892, \'job\': \'Screenplay\', \'name\': \'Joel Cohen\', \'profile_path\': \'/dAubAiZcvKFbboWlj7oXOkZnTSu.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f61\', \'department\': \'Writing\', \'gender\': 0, \'id\': 12893, \'job\': \'Screenplay\', \'name\': \'A

In [54]:
def get_director(obj):
    direct = []
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            direct.append(i['name'])
            break
    return direct

In [55]:
df['crew'] = df['crew'].apply(get_director)
df.head(2)

genres    id                       poster_path  \
0   [Animation, Comedy, Family]   862  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  [Adventure, Fantasy, Family]  8844  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
1        Jumanji  When siblings Judy and Peter discover an encha...   

                                production_companies  \
0                          [Pixar Animation Studios]   
1  [TriStar Pictures, Teitler Film, Interscope Co...   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   

                                             cast             crew  
0             [Tom Hanks, Tim Allen, Don Rickles]  [John Lasseter]  
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   [Joe Johnston]

Since, our final motive is to concatenate all the lists together to get tags, so we will also convert the overview column which is in string format to a list

In [56]:
df['overview'][0]

"Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences."

In [57]:
df['overview'] = df['overview'].apply(lambda x: x.split())
df.head(3)

genres     id                       poster_path  \
0   [Animation, Comedy, Family]    862  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  [Adventure, Fantasy, Family]   8844  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2             [Romance, Comedy]  15602  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   

     original_title                                           overview  \
0         Toy Story  [Led, by, Woody,, Andy's, toys, live, happily,...   
1           Jumanji  [When, siblings, Judy, and, Peter, discover, a...   
2  Grumpier Old Men  [A, family, wedding, reignites, the, ancient, ...   

                                production_companies  \
0                          [Pixar Animation Studios]   
1  [TriStar Pictures, Teitler Film, Interscope Co...   
2                     [Warner Bros., Lancaster Gate]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   
2  [fishing, best friend, duringcreditsstinger, o...   

                                             cast             crew  
0             [Tom Hanks, Tim Allen, Don Rickles]  [John Lasseter]  
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   [Joe Johnston]  
2      [Walter Matthau, Jack Lemmon, Ann-Margret]  [Howard Deutch]

Now we will concatenate all the lists to get a big paragraph which will act as our tag. But, before that we have to remove the space between the names as the first name and surname of many actors and directors might be same and our algorithm will mistake one for the other and give us wrong recommendations. So we can strip the white spaces and also make them lower case font to make them homogeneous.

In [58]:
space_cols = ['genres','production_companies','keywords','cast','crew']
for col in space_cols:
    df[col] = df[col].apply(lambda x: [i.replace(" ","") for i in x])

In [59]:
df.head(3)

genres     id                       poster_path  \
0   [Animation, Comedy, Family]    862  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  [Adventure, Fantasy, Family]   8844  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2             [Romance, Comedy]  15602  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   

     original_title                                           overview  \
0         Toy Story  [Led, by, Woody,, Andy's, toys, live, happily,...   
1           Jumanji  [When, siblings, Judy, and, Peter, discover, a...   
2  Grumpier Old Men  [A, family, wedding, reignites, the, ancient, ...   

                                production_companies  \
0                            [PixarAnimationStudios]   
1  [TriStarPictures, TeitlerFilm, InterscopeCommu...   
2                       [WarnerBros., LancasterGate]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [boardgame, disappearance, basedonchildren'sbo...   
2  [fishing, bestfriend, duringcreditsstinger, ol...   

                                          cast            crew  
0             [TomHanks, TimAllen, DonRickles]  [JohnLasseter]  
1  [RobinWilliams, JonathanHyde, KirstenDunst]   [JoeJohnston]  
2     [WalterMatthau, JackLemmon, Ann-Margret]  [HowardDeutch]

In [60]:
df['tags'] = df['genres'] + df['overview'] + df['production_companies'] + df['keywords'] + df['cast'] + df['crew']
df.head(3)

genres     id                       poster_path  \
0   [Animation, Comedy, Family]    862  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  [Adventure, Fantasy, Family]   8844  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2             [Romance, Comedy]  15602  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   

     original_title                                           overview  \
0         Toy Story  [Led, by, Woody,, Andy's, toys, live, happily,...   
1           Jumanji  [When, siblings, Judy, and, Peter, discover, a...   
2  Grumpier Old Men  [A, family, wedding, reignites, the, ancient, ...   

                                production_companies  \
0                            [PixarAnimationStudios]   
1  [TriStarPictures, TeitlerFilm, InterscopeCommu...   
2                       [WarnerBros., LancasterGate]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [boardgame, disappearance, basedonchildren'sbo...   
2  [fishing, bestfriend, duringcreditsstinger, ol...   

                                          cast            crew  \
0             [TomHanks, TimAllen, DonRickles]  [JohnLasseter]   
1  [RobinWilliams, JonathanHyde, KirstenDunst]   [JoeJohnston]   
2     [WalterMatthau, JackLemmon, Ann-Margret]  [HowardDeutch]   

                                                tags  
0  [Animation, Comedy, Family, Led, by, Woody,, A...  
1  [Adventure, Fantasy, Family, When, siblings, J...  
2  [Romance, Comedy, A, family, wedding, reignite...

In [61]:
final_df = df[['id','original_title','poster_path','tags']]
final_df.rename(columns={"original_title":"title"}, inplace=True)
final_df.head(3)

id             title                       poster_path  \
0    862         Toy Story  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1   8844           Jumanji  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  15602  Grumpier Old Men  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   

                                                tags  
0  [Animation, Comedy, Family, Led, by, Woody,, A...  
1  [Adventure, Fantasy, Family, When, siblings, J...  
2  [Romance, Comedy, A, family, wedding, reignite...

Now we have to convert the list entries in tags column into string entries.

In [62]:
final_df['tags'] = final_df['tags'].apply(lambda x: " ".join(x))
final_df.head(2)

id      title                       poster_path  \
0   862  Toy Story  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  8844    Jumanji  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

                                                tags  
0  Animation Comedy Family Led by Woody, Andy's t...  
1  Adventure Fantasy Family When siblings Judy an...

In [63]:
final_df['tags'][0]

"Animation Comedy Family Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences. PixarAnimationStudios jealousy toy boy friendship friends rivalry boynextdoor newtoy toycomestolife TomHanks TimAllen DonRickles JohnLasseter"

In [64]:
final_df['tags'] = final_df['tags'].apply(lambda x:x.lower())
final_df.head(3)

id             title                       poster_path  \
0    862         Toy Story  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1   8844           Jumanji  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  15602  Grumpier Old Men  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   

                                                tags  
0  animation comedy family led by woody, andy's t...  
1  adventure fantasy family when siblings judy an...  
2  romance comedy a family wedding reignites the ...

In [65]:
final_df.shape

(4763, 4)

We have to perform text vectorization, where similarity will be based on tags. We will convert the whole tags text in vector format and plot on a graph, all the different movies can be represented by vectors and the closeness is based on cosine distance.

We will concatenate all the tags to get a huge string and we will get top 10000 most common words. Then this will become our columns and all the individual movie tags will be our rows and the number of occurences of the given word in a particular tag will be entered in a row wise manner, where each row represents a vector of dimension **5,000*1** belonging to first movie. Likewise, we have in total 4763 movies so we get a table of **4763*5,000** dimensions. Plotting all these 4763 vectors in 5000 dimensional space and then fetching all the closest vectors to a given vector will give us the most closely associated movie to given movie.

We will not consider stop words during vectorization, which don't contribute in the sentence meaning. stop_words paramter will take care of this in the CountVectorizer class.

We see that many same words are repeated in different forms for example, adapt, adaptation and adapted are all same words. So, we need to get rid of this issue, we use ***stemming***.

In [66]:
#!pip install nltk

In [67]:
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [68]:
# Create a function stem which performs Stemming and gives us the root words
def stem_words(text):
    lst = []
    for i in text.split():
        lst.append(ps.stem(i))
    return " ".join(lst)

In [69]:
final_df['tags'] = final_df['tags'].apply(stem_words)
final_df['tags']

0       anim comedi famili led by woody, andy' toy liv...
1       adventur fantasi famili when sibl judi and pet...
2       romanc comedi a famili wed reignit the ancient...
3       comedi drama romanc cheat on, mistreat and ste...
4       comedi just when georg bank ha recov from hi d...
                              ...                        
4758    thriller horror romanc sciencefict the calv of...
4759    drama a griev doctor is be contact by hi late ...
4760    drama fantasi horror lestat de lioncourt is aw...
4761    drama comedi vietnam veteran leon barlow is st...
4762    action adventur drama a tale about vietnames r...
Name: tags, Length: 4763, dtype: object

### Vectorization

In [70]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(max_features=5000, stop_words='english')
vectors = cv.fit_transform(final_df['tags']).toarray()

In [71]:
cv.get_feature_names()

['000',
 '10',
 '100',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '18th',
 '18thcenturi',
 '1920',
 '1920s',
 '1930',
 '1940',
 '1944',
 '1950',
 '1950s',
 '1955',
 '1957',
 '1960',
 '1960s',
 '1965',
 '1970',
 '1976',
 '1980',
 '1980s',
 '1990',
 '1996',
 '1997',
 '1999',
 '19th',
 '19thcenturi',
 '20',
 '2000',
 '20th',
 '24',
 '25',
 '30',
 '3d',
 '40',
 '40acres',
 '50',
 '60',
 '80',
 '90',
 'abandon',
 'abc',
 'abduct',
 'abelferrara',
 'abil',
 'abl',
 'aboard',
 'abov',
 'abroad',
 'abruptli',
 'abus',
 'academ',
 'academi',
 'academy',
 'accept',
 'access',
 'accid',
 'accident',
 'acclaim',
 'accompani',
 'accomplish',
 'account',
 'accus',
 'ace',
 'achiev',
 'acid',
 'acquaint',
 'acquir',
 'act',
 'action',
 'activ',
 'activist',
 'actor',
 'actress',
 'actual',
 'ad',
 'ada',
 'adam',
 'adamsandl',
 'adapt',
 'add',
 'addict',
 'addit',
 'adjust',
 'admir',
 'admit',
 'adolesc',
 'adopt',
 'ador',
 'adrianlyn',
 'adrienbrodi',
 'adult',
 'adultanim',
 'adult

In [72]:
vectors.shape

(4763, 5000)

In such high dimensional data, Euclidean distance fails and it is not a good measure to calculate the closest distance, instead we use a more reliable measure called cosine distance which is prsent in scikit learn.

In [73]:
from sklearn.metrics.pairwise import cosine_similarity

In [74]:
#type(vectors)
#vaex_vectors = vaex.from_arrays()
#vaex_vectors

In [75]:
similarity = cosine_similarity(vectors)

In [76]:
# it gives similarity of index 1 movie with all others. We see that at second position value is 1 because similarity with itself
similarity[1]

array([0.05416186, 1.        , 0.06271815, ..., 0.01580349, 0.        ,
       0.02001602])

For getting top 10 closest movies to any given movie we have to sort the above obtained similarity array, but in this way we also loose the index positions of corresponding movies. So, we use enumerate to preserve this index position of sorted array.

In [77]:
# we get the index positon and corresponding values
# now we want to sort but not on the basis of index positon but on the similarity metric, for this we use key=lambda x:x[1]
similarity_tuple = sorted(list(enumerate(similarity[1])), reverse=True, key=lambda x:x[1])
similarity_tuple

[(1, 1.0000000000000004),
 (3688, 0.2377286555250982),
 (56, 0.22188007849009167),
 (2362, 0.18681617943926834),
 (1387, 0.17755520605710873),
 (557, 0.17650452162436564),
 (1873, 0.17118419700436519),
 (521, 0.16647802295407294),
 (3205, 0.16531163063339516),
 (4675, 0.1640825308284734),
 (8, 0.16343011261515336),
 (2906, 0.16182267881254697),
 (1833, 0.15907119074394444),
 (2093, 0.15848577035006547),
 (4741, 0.15724272550828775),
 (4338, 0.15480470613460084),
 (1426, 0.15348515337873014),
 (4619, 0.14824986333222026),
 (3985, 0.147087101353638),
 (1830, 0.1444316289348238),
 (4275, 0.1441153384245784),
 (4656, 0.1415948439609786),
 (2887, 0.14153462926807447),
 (619, 0.1386750490563073),
 (1946, 0.1386750490563073),
 (3773, 0.13867504905630726),
 (2174, 0.13728129459672883),
 (1207, 0.13678822577626598),
 (203, 0.13678822577626595),
 (3795, 0.13617619485048055),
 (1847, 0.1333393806079054),
 (1920, 0.1315587028960544),
 (3122, 0.1307440900921227),
 (2513, 0.1287565506561512),
 (2816

In [78]:
similarity_tuple[1:6]

[(3688, 0.2377286555250982),
 (56, 0.22188007849009167),
 (2362, 0.18681617943926834),
 (1387, 0.17755520605710873),
 (557, 0.17650452162436564)]

In [79]:
final_df.iloc[4453].title

'The Little Foxes'

In [80]:
# getting the index position of the movie
final_df[final_df['title'] == 'Jumanji'].index[0]

1

In [81]:
final_df[final_df['title'] == "Harry Potter and the Philosopher's Stone"]

id                                     title  \
4542  671  Harry Potter and the Philosopher's Stone   

                           poster_path  \
4542  /dCtFvscYcXQKTNvyyaQr2g2UacJ.jpg   

                                                   tags  
4542  adventur fantasi famili harri potter ha live u...

In [92]:
final_df[final_df['title'] == "Richard Pryor: Here and Now"]

id                        title                       poster_path  \
4733  21921  Richard Pryor: Here and Now  /7tBphDdBr6Ws5pmCmPllW3m03Z1.jpg   

                                                   tags  
4733  comedi part live stand-up performance, part do...

In [96]:
# this doesn't have poster_path at tmdb api
final_df[final_df['title'] == "All Access: Front Row. Backstage. Live!"]

id                                    title  \
4116  103489  All Access: Front Row. Backstage. Live!   

                           poster_path                                    tags  
4116  /eRnLo1ACCrFq4P6xjPTp5IZpf8F.jpg  front row. backstage. live! imax film.

In [82]:
final_df.tail()

id                title                       poster_path  \
4758  43715    The Deadly Mantis  /k7iUnaelPU5kwEEVZV7xW7fvFIS.jpg   
4759  10052            Dragonfly  /gs3fFBoiEaExy5x3m5yAQpmVhXs.jpg   
4760  11979  Queen of the Damned  /jSNMRMPkXUqcqWm0xUPUviRfWfF.jpg   
4761  75151         Big Bad Love   /iDkgvzFjIvs7n0C4gDgAh7hnSY.jpg   
4762  35868         Green Dragon  /4qNxipsoqADZu5ZACK6uJv4qkWV.jpg   

                                                   tags  
4758  thriller horror romanc sciencefict the calv of...  
4759  drama a griev doctor is be contact by hi late ...  
4760  drama fantasi horror lestat de lioncourt is aw...  
4761  drama comedi vietnam veteran leon barlow is st...  
4762  action adventur drama a tale about vietnames r...

In [83]:
final_df.shape

(4763, 4)

In [84]:
def recommend_top_movies(movie):
    movie_index = final_df[final_df['title'] == movie].index[0]
    distances = similarity[movie_index]
    movies_tuple = sorted(list(enumerate(distances)), reverse=True, key=lambda x:x[1])[1:6]
    
    for entry in movies_tuple:
        print(final_df.iloc[entry[0]].title)

In [85]:
recommend_top_movies('Jumanji')

Dungeons & Dragons
The Indian in the Cupboard
eXistenZ
The Second Jungle Book: Mowgli & Baloo
Aladdin


In [86]:
recommend_top_movies('Toy Story')

Toy Story 2
Brighton Beach Memoirs
The Adventures of Elmo in Grouchland
Carried Away
The Million Dollar Duck


In [87]:
recommend_top_movies("Harry Potter and the Philosopher's Stone")

The Adventures of Elmo in Grouchland
Song of the South
Le colonel Chabert
Commandments
Carried Away


In [91]:
recommend_top_movies("Richard Pryor: Here and Now")

Martin Lawrence: You So Crazy
Punchline
All Access: Front Row. Backstage. Live!
The American Astronaut
The Original Kings of Comedy


### Dumping file

In [88]:
import pickle

pickle.dump(final_df.to_dict(), open('movies.pkl', 'wb'))

In [89]:
#compressed_pickle('similarity_compressed', similarity)

In [90]:
pickle.dump(similarity, open('similarity_scores.pkl','wb'))